gcstorage/gcstorage.go

```diff
@@ -0,0 +1,253 @@
+package gcstorage
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/goware/cachestore"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+var ErrNotSupported = fmt.Errorf("not supported")
+
+type cacheObject[T any] struct {
+	Object    T         `json:"object"`
+	ExpiresAt time.Time `json:"expires_at"`
+}
+
+type GCStorage[V any] struct {
+	keyPrefix string
+
+	defaultKeyExpiry time.Duration
+
+	client       *storage.Client
+	bucketHandle *storage.BucketHandle
+}
+
+func Backend(cfg *Config, opts ...cachestore.StoreOptions) cachestore.Backend {
+	options := cachestore.ApplyOptions(opts...)
+	cfg.StoreOptions = options
+	return cfg
+}
+
+func NewWithBackend[V any](backend cachestore.Backend, opts ...cachestore.StoreOptions) (cachestore.Store[V], error) {
+	cfg, ok := backend.(*Config)
+	if !ok {
+		return nil, fmt.Errorf("cachestore/redis: invalid backend config supplied")
+	}
+	for _, opt := range opts {
+		opt.Apply(&cfg.StoreOptions)
+	}
+	return New[V](cfg, cfg.StoreOptions)
+}
+
+func New[V any](cfg *Config, opts ...cachestore.StoreOptions) (cachestore.Store[V], error) {
+	for _, opt := range opts {
+		opt.Apply(&cfg.StoreOptions)
+	}
+
+	var gcOpts []option.ClientOption
+	if cfg.Credentials != nil {
+		gcOpts = append(gcOpts, option.WithCredentials(cfg.Credentials))
+	}
+
+	client, err := storage.NewClient(context.Background(), gcOpts...)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
We should pass ctx key to client on New.
```suggestion
func NewWithBackend[V any](ctx context.Context, backend cachestore.Backend, opts ...cachestore.StoreOptions) (cachestore.Store[V], error) {
	cfg, ok := backend.(*Config)
	if !ok {
		return nil, fmt.Errorf("cachestore/redis: invalid backend config supplied")
	}
	for _, opt := range opts {
		opt.Apply(&cfg.StoreOptions)
	}
	return New[V](ctx, cfg, cfg.StoreOptions)
}

func New[V any](ctx context.Context, cfg *Config, opts ...cachestore.StoreOptions) (cachestore.Store[V], error) {
	for _, opt := range opts {
		opt.Apply(&cfg.StoreOptions)
	}

	var gcOpts []option.ClientOption
	if cfg.Credentials != nil {
		gcOpts = append(gcOpts, option.WithCredentials(cfg.Credentials))
	}

	client, err := storage.NewClient(ctx, gcOpts...)
	if err != nil {
		return nil, err
	}

```

Reply:
I can not because it would be inconsistent with other implementations and ```cachestoragectl```. The context in here is used to extract configuration / http client. I haven't found anything that would make an api call. 

We could either leave it as it is or alter all implementations. WDYT?

Reply:
let's leave it

btw: I don't think the whole backend should receive `ctx`.. instead, we should pass it to individual i/o methods, which we do